### PR TITLE
Clean up RoutingNodes

### PR DESCRIPTION
**Pull request by ywelsch**
This PR contains a number of cleanups related to recent changes in RoutingNodes:
- PR #17821 (Immutable ShardRouting) changed RoutingNode to use a map indexed by ShardId to manage ShardRouting elements. This means that we can directly select the right ShardRouting without iterating over all elements. This lets us get rid of RoutingNodeIterator and all kind of iterations all over the place.
- Second cleanup is an extension of #18390 (Expose cluster state before reroute in RoutingAllocation instead of RoutingNodes). We should not reexpose RoutingTable in RoutingNodes and only use it in the constructor. This makes it clear that the RoutingTable is only used to construct the RoutingNodes and can diverge from it afterwards (only RoutingNodes is mutable).
- Remove AllocationService.applyNewNodes() (that is already done as part of construction of RoutingNodes)

### CODE REVIEW

**Comment by ywelsch** [comment]
@s1monw can I interest you in reviewing this? :-)

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java [hunk 1]
```diff
@@ -119,6 +115,11 @@ void update(ShardRouting oldShard, ShardRouting newShard) {
         assert previousValue == oldShard : "expected shard " + previousValue + " but was " + oldShard;
     }
 
+    void remove(ShardRouting shard) {
+        ShardRouting previousValue = shards.remove(shard.shardId());
+        assert previousValue == shard : "expected shard " + previousValue + " but was " + shard;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java]: should this be a hard check?

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java [hunk 2]
```diff
@@ -303,19 +302,11 @@ public ShardRouting activeReplica(ShardRouting shard) {
...
-        return assignedShards(shard.shardId());
-    }
-
-    /**
      * Returns <code>true</code> iff all replicas are active for the given shard routing. Otherwise <code>false</code>
      */
-    public boolean allReplicasActive(ShardRouting shardRouting) {
-        final List<ShardRouting> shards = assignedShards(shardRouting.shardId());
-        if (shards.isEmpty() || shards.size() < this.routingTable.index(shardRouting.index().getName()).shard(shardRouting.id()).size()) {
+    public boolean allReplicasActive(ShardId shardId, int expectedAssignedShardsCount) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java]: explain the 2nd arg in the javadocs?

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java [hunk 3]
```diff
@@ -480,32 +462,25 @@ private boolean applyStartedShards(RoutingNodes routingNodes, Iterable<? extends
...
-                while (sourceRoutingNode.hasNext()) {
-                    ShardRouting shard = sourceRoutingNode.next();
-                    if (shard.isRelocationSourceOf(startedShard)) {
-                        dirty = true;
-                        sourceRoutingNode.remove();
-                        break;
+                    if (startedShard.relocatingNodeId() != null) {
+                        // relocation target has been started, remove relocation source
+                        RoutingNode relocationSourceNode = routingNodes.node(startedShard.relocatingNodeId());
+                        if (relocationSourceNode != null) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: hmm how can this be null? is this due to sending shard started multiple times?

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java [hunk 4]
```diff
@@ -177,7 +180,7 @@ public AllocationDeciders deciders() {
      * @return current routing table
      */
     public RoutingTable routingTable() {
-        return routingNodes.routingTable();
+        return routingTable;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java]: good!

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/CancelAllocationCommand.java [hunk 5]
```diff
@@ -120,86 +122,80 @@ public boolean allowPrimary() {
...
-        boolean found = false;
-        for (RoutingNodes.RoutingNodeIterator it = allocation.routingNodes().routingNodeIter(discoNode.getId()); it.hasNext(); ) {
-            ShardRouting shardRouting = it.next();
-            if (!shardRouting.shardId().getIndex().getName().equals(index)) {
-                continue;
+        ShardRouting shardRouting = null;
+        RoutingNodes routingNodes = allocation.routingNodes();
+        RoutingNode routingNode = routingNodes.node(discoNode.getId());
+        if (routingNode != null) {
+            IndexMetaData indexMetaData = allocation.metaData().index(index());
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/CancelAllocationCommand.java]: should we throw IndexNotFound here?

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/CancelAllocationCommand.java [hunk 6]
```diff
@@ -120,86 +122,80 @@ public boolean allowPrimary() {
...
                     if (explain) {
                         return new RerouteExplanation(this, allocation.decision(Decision.NO, "cancel_allocation_command",
-                                "can't cancel " + shardId + " on node " + discoNode + ", shard is primary and started"));
+                                "can't cancel " + shardId + " on node " + discoNode + ", shard is primary and initializing its state"));
                     }
                     throw new IllegalArgumentException("[cancel_allocation] can't cancel " + shardId + " on node " +
-                            discoNode + ", shard is primary and started");
+                            discoNode + ", shard is primary and initializing its state");
+                }
+                boolean moveToUnassigned = true;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/CancelAllocationCommand.java]: any chance we can shard this code with AllocationService

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RebalanceOnlyWhenActiveAllocationDecider.java [hunk 7]
```diff
@@ -40,7 +40,8 @@ public RebalanceOnlyWhenActiveAllocationDecider(Settings settings) {
     public Decision canRebalance(ShardRouting shardRouting, RoutingAllocation allocation) {
         // its ok to check for active here, since in relocation, a shard is split into two in routing
         // nodes, once relocating, and one initializing
-        if (!allocation.routingNodes().allReplicasActive(shardRouting)) {
+        int expectedAssignedShardsCount = allocation.routingTable().index(shardRouting.index().getName()).shard(shardRouting.id()).size();
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RebalanceOnlyWhenActiveAllocationDecider.java]: not sure if that is an improvement to be honest? it seems like it's way more work to get this outside of the method? maybe have an overloaded version?

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalanceConfigurationTests.java [hunk 8]
```diff
@@ -258,26 +257,6 @@ private void assertIndexBalance(ESLogger logger, RoutingNodes nodes, int numberO
         }
     }
 
-    private void assertPrimaryBalance(ESLogger logger, RoutingNodes nodes, int numberOfNodes, int numberOfIndices, int numberOfReplicas, int numberOfShards, float treshold) {
```
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalanceConfigurationTests.java]: dead code?

Comment (s1monw) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalanceConfigurationTests.java]: yeah I must have missed that one

**Comment by s1monw** [comment]
left small comments - look awesome

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java (hunk 1 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java]: I prefer this to be a soft check as it only verifies if we're using the internal API properly.

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java (hunk 2 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java]: I rewrote the method, relying on index metadata to get the number of shards in the replica group instead.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java (hunk 3 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: I cannot think of a situation where it is null. At that point in the method, startedShard is the existing entry in the routing table. So it should also have the corresponding entry for the relocation source. I've omitted the null checks...

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/CancelAllocationCommand.java (hunk 5 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/CancelAllocationCommand.java]: agree

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/CancelAllocationCommand.java (hunk 6 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/CancelAllocationCommand.java]: I gave it a try, have another look and see if you like it.

**Comment by ywelsch** [comment]
@s1monw I've updated the PR with your suggestions. Please have another look.

File: core/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java]: ok cool

**Comment by s1monw** [comment]
LGTM thanks for the cleanups

**Pull request merged**